internal/cloud/realm/app.go

```diff
@@ -0,0 +1,137 @@
+package realm
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/10gen/realm-cli/internal/utils/api"
+)
+
+// UserProfile contains all of the roles associated with a user
+type UserProfile struct {
+	Roles []Role `json:"roles"`
+}
+
+// Role contains a GrouID field that maps to a project ID
+type Role struct {
+	GroupID string `json:"group_id"`
+}
+
+// App represents basic Realm App data
+type App struct {
+	ID          string `json:"_id"`
+	GroupID     string `json:"group_id"`
+	ClientAppID string `json:"client_app_id"`
+	Name        string `json:"name"`
+}
+
+var (
+	errGroupNotFound = errors.New("group could not be found")
+)
+
+var (
+	appsByGroupIDEndpoint = adminAPI + "/groups/%s/apps"
+	userProfileEndpoint   = adminAPI + "/auth/profile"
+)
+
+// FilterGroupIDsFromUserProfile returns all available group ids for a given user
+func (pd *UserProfile) FilterGroupIDsFromUserProfile(projectFlag string) []string {
+	groupIDs := []string{}
+	set := make(map[string]bool)
+	for _, role := range pd.Roles {
+		// Break early if we find the groupID that matches the project flag
+		if len(projectFlag) == 0 && !set[role.GroupID] {
+			groupIDs = append(groupIDs, role.GroupID)
+			set[role.GroupID] = true
+			continue
+		}
+
+		if role.GroupID == projectFlag {
+			return []string{role.GroupID}
+		}
+	}
+
+	return groupIDs
+}
+
+func (c *client) GetUserProfile() (UserProfile, error) {
+	opts, _ := api.JSONRequestOptions(nil)
+
+	res, resErr := c.do(http.MethodGet, appsByGroupIDEndpoint, opts)
+	if resErr != nil {
+		return UserProfile{}, resErr
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return UserProfile{}, UnmarshalServerError(res)
+	}
+
+	dec := json.NewDecoder(res.Body)
+	defer res.Body.Close()
+
+	var profile UserProfile
+	if err := dec.Decode(&profile); err != nil {
+		return UserProfile{}, err
+	}
+
+	return profile, nil
+}
+
+func (c *client) FindProjectAppByClientAppID(groupIDs []string, appFlag string) ([]App, error) {
+	var appArr []App
+	appArr = make([]App, 0)
+	for _, groupID := range groupIDs {
+		apps, err := c.FetchAppsByGroupID(groupID)
+		if err != nil && err != errGroupNotFound {
+			//TODO REALMC-7156: should we fail silently or hard here
```

Comment:
I had a silent failure here in the scenario where we try to get apps from one of the project IDs tied to the user and one of the calls fails. Here we just keep going as if nothing went wrong as long as the error is not a GroupNotFound error. I am fine letting all errors go through too. I copied this logic from the Realm CLI but in our use cases, I don't think it matters why we failed to fetch the apps for a group ID

Reply:
Changed to fail hard, as per offline discussion + other comment